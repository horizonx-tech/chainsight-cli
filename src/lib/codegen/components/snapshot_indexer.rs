use std::collections::HashMap;

use proc_macro2::TokenStream;
use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::{
    lib::codegen::{canisters, components::common::SourceType, scripts},
    types::{ComponentType, Network},
};

use super::{
    algorithm_lens::LensTargets,
    common::{
        custom_tags_interval_sec, ComponentManifest, ComponentMetadata, Datasource,
        DestinationType, Sources,
    },
};

/// Component Manifest: Snapshot
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct SnapshotIndexerComponentManifest {
    pub version: String,
    pub metadata: ComponentMetadata,
    pub datasource: Datasource,
    pub storage: SnapshotStorage,
    pub interval: u32,
    pub lens_targets: Option<LensTargets>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct LensTarget {
    pub identifiers: Vec<String>,
}

impl SnapshotIndexerComponentManifest {
    pub fn new(
        label: &str,
        description: &str,
        version: &str,
        datasource: Datasource,
        storage: SnapshotStorage,
        interval: u32,
    ) -> Self {
        Self {
            version: version.to_owned(),
            metadata: ComponentMetadata {
                label: label.to_owned(),
                type_: ComponentType::SnapshotIndexer,
                description: description.to_owned(),
                tags: Some(vec![
                    "ERC-20".to_string(),
                    "Ethereum".to_string(),
                    "DAI".to_string(),
                ]),
            },
            datasource,
            storage,
            interval,
            lens_targets: None,
        }
    }
}
impl ComponentManifest for SnapshotIndexerComponentManifest {
    fn to_str_as_yaml(&self) -> anyhow::Result<String> {
        let yaml = serde_yaml::to_string(&self)?;
        Ok(self.yaml_str_with_configs(yaml, "snapshot_indexer".to_string()))
    }

    fn validate_manifest(&self) -> anyhow::Result<()> {
        canisters::snapshot_indexer::validate_manifest(self)
    }

    fn generate_codes(
        &self,
        _interface_contract: Option<ethabi::Contract>,
    ) -> anyhow::Result<TokenStream> {
        canisters::snapshot_indexer::generate_codes(self)
    }

    fn generate_scripts(&self, network: Network) -> anyhow::Result<String> {
        scripts::snapshot_indexer::generate_scripts(self, network)
    }

    fn component_type(&self) -> ComponentType {
        ComponentType::SnapshotIndexer
    }

    fn metadata(&self) -> &ComponentMetadata {
        &self.metadata
    }

    fn destination_type(&self) -> Option<DestinationType> {
        None
    }

    fn required_interface(&self) -> Option<String> {
        self.datasource.method.interface.clone()
    }
    fn user_impl_required(&self) -> bool {
        true
    }
    fn generate_user_impl_template(&self) -> anyhow::Result<TokenStream> {
        canisters::snapshot_indexer::generate_app(self)
    }
    fn get_sources(&self) -> Sources {
        let mut attr = HashMap::new();
        let mut method_identifier = self.datasource.clone().method.identifier;
        if method_identifier.contains(':') {
            method_identifier = method_identifier.split(':').collect::<Vec<&str>>()[0]
                .to_string()
                .replace(' ', "")
                .replace("()", "");
        }
        if self.lens_targets.is_some() {
            let targets = self.lens_targets.clone().unwrap().identifiers;
            attr.insert("sources".to_string(), json!(targets));
        }

        attr.insert("function_name".to_string(), json!(method_identifier));
        Sources {
            source: self.datasource.location.id.clone(),
            source_type: SourceType::Chainsight,
            attributes: attr,
        }
    }
    fn custom_tags(&self) -> HashMap<String, String> {
        let mut res = HashMap::new();
        let (interval_key, interval_val) = custom_tags_interval_sec(self.interval);
        res.insert(interval_key, interval_val);
        res
    }
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
pub struct SnapshotStorage {
    pub with_timestamp: bool,
}
impl SnapshotStorage {
    pub fn new(with_timestamp: bool) -> Self {
        Self { with_timestamp }
    }
}
impl Default for SnapshotStorage {
    fn default() -> Self {
        Self::new(true)
    }
}

#[cfg(test)]
mod tests {
    use jsonschema::JSONSchema;

    use crate::lib::codegen::components::common::{
        CanisterIdType, DatasourceLocation, DatasourceMethod, DatasourceType,
    };

    use super::*;

    #[test]
    fn test_to_manifest_struct_for_chain() {
        let yaml = r#"
version: v1
metadata:
    label: sample_snapshot_indexer_chain
    type: snapshot_indexer
    description: Description
    tags:
    - ERC-20
    - Ethereum
datasource:
    type: contract
    location:
        id: 6b175474e89094c44da98b954eedeac495271d0f
        args:
            network_id: 1
            rpc_url: https://mainnet.infura.io/v3/<YOUR_KEY>
    method:
        identifier: totalSupply():(uint256)
        interface: ERC20.json
        args: []
storage:
    with_timestamp: true
interval: 3600
        "#;

        let result = serde_yaml::from_str::<SnapshotIndexerComponentManifest>(yaml);
        assert!(result.is_ok());
        let component = result.unwrap();
        assert_eq!(
            component,
            SnapshotIndexerComponentManifest {
                version: "v1".to_owned(),
                metadata: ComponentMetadata {
                    label: "sample_snapshot_indexer_chain".to_owned(),
                    type_: ComponentType::SnapshotIndexer,
                    description: "Description".to_string(),
                    tags: Some(vec!["ERC-20".to_string(), "Ethereum".to_string()])
                },
                datasource: Datasource {
                    type_: DatasourceType::Contract,
                    location: DatasourceLocation::new_contract(
                        "6b175474e89094c44da98b954eedeac495271d0f".to_string(),
                        1,
                        "https://mainnet.infura.io/v3/<YOUR_KEY>".to_string(),
                    ),
                    method: DatasourceMethod {
                        identifier: "totalSupply():(uint256)".to_owned(),
                        interface: Some("ERC20.json".to_string()),
                        args: vec![]
                    }
                },
                storage: SnapshotStorage {
                    with_timestamp: true,
                },
                lens_targets: None,
                interval: 3600
            }
        );
    }

    #[test]
    fn test_to_manifest_struct_for_icp() {
        let yaml = r#"
version: v1
metadata:
    label: sample_snapshot_indexer_icp
    type: snapshot_indexer
    description: Description
    tags:
    - ERC-20
    - Ethereum
datasource:
    type: canister
    location:
        id: datasource_canister_id
        args:
            id_type: canister_name
    method:
        identifier: 'get_last_snapshot : () -> (record { value : text; timestamp : nat64 })'
        args: []
storage:
    with_timestamp: true
interval: 3600
        "#;

        let result = serde_yaml::from_str::<SnapshotIndexerComponentManifest>(yaml);
        assert!(result.is_ok());
        let component = result.unwrap();
        assert_eq!(
            component,
            SnapshotIndexerComponentManifest {
                version: "v1".to_owned(),
                metadata: ComponentMetadata {
                    label: "sample_snapshot_indexer_icp".to_owned(),
                    type_: ComponentType::SnapshotIndexer,
                    description: "Description".to_string(),
                    tags: Some(vec!["ERC-20".to_string(), "Ethereum".to_string()])
                },
                datasource: Datasource {
                    type_: DatasourceType::Canister,
                    location: DatasourceLocation::new_canister(
                        "datasource_canister_id".to_string(),
                        CanisterIdType::CanisterName
                    ),
                    method: DatasourceMethod {
                        identifier:
                            "get_last_snapshot : () -> (record { value : text; timestamp : nat64 })"
                                .to_owned(),
                        interface: None,
                        args: vec![]
                    }
                },
                storage: SnapshotStorage {
                    with_timestamp: true,
                },
                lens_targets: None,
                interval: 3600
            }
        );
        let schema = serde_json::from_str(include_str!(
            "../../../../resources/schema/snapshot_indexer.json"
        ))
        .expect("Invalid json");
        let instance = serde_yaml::from_str(yaml).expect("Invalid yaml");
        let compiled = JSONSchema::compile(&schema).expect("Invalid schema");
        let result = compiled.validate(&instance);
        assert!(result.is_ok());
    }
}
