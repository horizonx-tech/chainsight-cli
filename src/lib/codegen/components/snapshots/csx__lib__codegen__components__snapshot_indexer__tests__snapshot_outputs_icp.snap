---
source: src/lib/codegen/components/snapshot_indexer.rs
expression: "SrcString::from(&manifest.generate_codes(Option::None).unwrap())"
---
use candid::{
Decode, Encode };
use chainsight_cdk_macros::{
init_in, manage_single_state, setup_func, prepare_stable_structure, stable_memory_for_vec, StableMemoryStorable, timer_task_func, chainsight_common, did_export, snapshot_icp_source };
use chainsight_cdk::rpc::{
CallProvider, Caller, Message };
init_in!();
chainsight_common!(3600);
manage_single_state!("target_canister", String, false);
setup_func!({
target_canister:String
}
;
prepare_stable_structure!();
stable_memory_for_vec!("snapshot", Snapshot, 0, true);
timer_task_func!("set_task", "execute_task", true);
#[derive(Clone, Debug, candid::CandidType, candid::Deserialize, serde::Serialize, StableMemoryStorable)]
#[stable_mem_storable_opts(max_size = 10000, is_fixed_size = false)]
pub struct Snapshot {
pub value:SnapshotValue, pub timestamp:u64,
}
type SnapshotValue = CustomResponseStruct;
fn _get_last_snapshot_value() -> SnapshotValue {
get_last_snapshot().value
}
fn _get_top_snapshot_values(n:u64) -> Vec<SnapshotValue>{
get_top_snapshots(n).iter().map(| s | s.value.clone()).collect()
}
fn _get_snapshot_value(idx:u64) -> SnapshotValue {
get_snapshot(idx).value
}
#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_last_snapshot_value() -> SnapshotValue {
_get_last_snapshot_value()
}
#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_top_snapshot_values(n:u64) -> Vec<SnapshotValue>{
_get_top_snapshot_values(n)
}
#[ic_cdk::query]
#[candid::candid_method(query)]
pub fn get_snapshot_value(idx:u64) -> SnapshotValue {
_get_snapshot_value(idx)
}
#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_last_snapshot_value(input:std::vec::Vec<u8>) -> std::vec::Vec<u8>{
use chainsight_cdk::rpc::Receiver;
chainsight_cdk::rpc::ReceiverProviderWithoutArgs::< SnapshotValue>:: new(proxy(), _get_last_snapshot_value, ).reply(input).await
}
#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_top_snapshot_values(input:std::vec::Vec<u8>) -> std::vec::Vec<u8>{
use chainsight_cdk::rpc::Receiver;
chainsight_cdk::rpc::ReceiverProvider::< u64, Vec<SnapshotValue>>::new(proxy(), _get_top_snapshot_values, ).reply(input).await
}
#[ic_cdk::update]
#[candid::candid_method(update)]
pub async fn proxy_get_snapshot_value(input:std::vec::Vec<u8>) -> std::vec::Vec<u8>{
use chainsight_cdk::rpc::Receiver;
chainsight_cdk::rpc::ReceiverProvider::< u64, SnapshotValue>:: new(proxy(), _get_snapshot_value, ).reply(input).await
}
#[derive(Clone, Debug, candid::CandidType, serde::Serialize, candid::Deserialize)]
pub struct CustomResponseStruct {
pub value:String, pub timestamp:u64
}
type CallCanisterArgs = sample_snapshot_indexer_icp::CallCanisterArgs;
pub fn call_args() -> CallCanisterArgs {
sample_snapshot_indexer_icp::call_args()
}
snapshot_icp_source!("proxy_get_last_snapshot");
type CallCanisterResponse = SnapshotValue;
async fn execute_task() {
let current_ts_sec = ic_cdk::api::time() / 1000000;
let target_canister = candid::Principal::from_text(get_target_canister()).unwrap();
let px = _get_target_proxy(target_canister).await;
let call_result = CallProvider::new().call(Message::new::< CallCanisterArgs>(call_args(), px.clone(), "proxy_get_last_snapshot").unwrap()).await;
if let Err(err) = call_result {
ic_cdk::println!("error: {
:?
}
, err);
return;

}
let res = call_result.unwrap().reply::< CallCanisterResponse>();
if let Err(err) = res {
ic_cdk::println!("error: {
:?
}
, err);
return;

}
let datum = Snapshot {
value:res.unwrap().clone(), timestamp:current_ts_sec, };
add_snapshot(datum.clone());
ic_cdk::println!("ts={

}
  value={
:?
}
, datum.timestamp, datum.value);

}
did_export!("sample_snapshot_indexer_icp");

